File: java/src/main/java/com/inariwatch/capture/Hmac.java
package com.inariwatch.capture;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;

public final class Hmac {
    private Hmac() {}

    /** HMAC-SHA256(payload, secret) -> lowercase hex. Matches the Node /
     *  Python / Go / Rust SDKs byte-for-byte. */
    public static String signSha256Hex(byte[] payload, String secret) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] sig = mac.doFinal(payload);
            StringBuilder hex = new StringBuilder(sig.length * 2);
            for (byte b : sig) hex.append(String.format("%02x", b));
            return hex.toString();
        } catch (Exception e) {
            throw new IllegalStateException("HMAC-SHA256 unavailable", e);
        }
    }
}

File: csharp/src/Hmac.cs
using System.Security.Cryptography;
using System.Text;

namespace InariWatch.Capture;

public static class Hmac
{
    /// <summary>
    /// HMAC-SHA256(payload, secret) -> lowercase hex. Matches the Node /
    /// Python / Go / Rust / Java SDKs byte-for-byte.
    /// </summary>
    public static string SignSha256Hex(byte[] payload, string secret)
    {
        using var mac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        byte[] sig = mac.ComputeHash(payload);
        var sb = new StringBuilder(sig.Length * 2);
        foreach (byte b in sig) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }
}
